integration/data/test_restore_to.py

```diff
@@ -0,0 +1,252 @@
+    output2_raw_backing = read_file(
+        output_raw_path, offset1, length0 - offset1)
+    assert output2_raw_snap2 == snap2_data
+    assert output2_raw_snap1 == snap1_data[offset2: length1]
+    assert output2_raw_backing == backing_data
+    os.remove(output_raw_path)
+    assert not os.path.exists(output_raw_path)
+
+    cmd.restore_to(backup, file(BACKING_FILE_QCOW2),
+                   output_qcow2_path, IMAGE_FORMAT_QCOW2)
+    output2_qcow2_snap2 = read_qcow2_file_without_backing_file(
+        output_qcow2_path, offset0, length2)
+    output2_qcow2_snap1 = read_qcow2_file_without_backing_file(
+        output_qcow2_path, offset2, length1 - length2)
+    output2_qcow2_backing = read_qcow2_file_without_backing_file(
+        output_qcow2_path, offset1, length0 - offset1)
+    assert output2_qcow2_snap2 == snap2_data
+    assert output2_qcow2_snap1 == snap1_data[offset2: length1]
+    assert output2_qcow2_backing == backing_data
+    os.remove(output_qcow2_path)
+    assert not os.path.exists(output_qcow2_path)
+
+    cmd.snapshot_revert(snap0)
+    rm_snaps(snap1, snap2)
+    rm_backups(backup)
+    check_backing(backing_dev)
+
+
+def restore_to_without_backing_file_test(dev, backup_target):
+    length0 = 256
+    length1 = 128
+    offset0 = 0
+    offset1 = length1 + offset0
+
+    output_raw_path = file(OUTPUT_FILE_RAW)
+    output_qcow2_path = file(OUTPUT_FILE_QCOW2)
+
+    # create 1 empty snapshot for converting to init state.
+    snap0 = cmd.snapshot_create()
+
+    # create 1 snapshot with 256B data.
+    # output = snap2(offset0, length1)
+    snap1_data = common.random_string(length0)
+    common.verify_data(dev, offset0, snap1_data)
+    snap1 = cmd.snapshot_create()
+    backup = create_backup(backup_target, snap1)
+
+    cmd.restore_to(backup, "",
+                   output_raw_path, IMAGE_FORMAT_RAW)
+    output1_raw = read_file(output_raw_path, offset0, length0)
+    assert output1_raw == snap1_data
+    os.remove(output_raw_path)
+    assert not os.path.exists(output_raw_path)
+
+    cmd.restore_to(backup, "",
+                   output_qcow2_path, IMAGE_FORMAT_QCOW2)
+    output1_qcow2 = read_qcow2_file_without_backing_file(
+        output_qcow2_path, offset0, length0)
+    assert output1_qcow2 == snap1_data
+    os.remove(output_qcow2_path)
+    assert not os.path.exists(output_qcow2_path)
+
+    cmd.snapshot_revert(snap0)
+    rm_snaps(snap1)
+    rm_backups(backup)
+
+    # create 2 snapshots with 256B data and 128B data
+    # output = snap2(offset0, length1 - length2) +
+    #          snap1(offset2, length2)
+    snap1_data = common.random_string(length0)
+    common.verify_data(dev, offset0, snap1_data)
+    snap1 = cmd.snapshot_create()
+    snap2_data = common.random_string(length1)
+    common.verify_data(dev, offset0, snap2_data)
+    snap2 = cmd.snapshot_create()
+    backup = create_backup(backup_target, snap2)
+
+    cmd.restore_to(backup, file(BACKING_FILE_QCOW2),
```

Comment:
This volume doesn't have a backing file, why add the backing file here? And below.